Signing: fatal error: concurrent map writes

**Describe the bug**

A `fatal error: concurrent map writes` is raised when signing binaries. This is not always the case. In 1/3 cases it fails.

I use gon to sign & notarize a binary for macos, the other os have an own build configuration. I used a sample similar to the one in the docs.

**To Reproduce**

Just run `goreleaser`

```yaml
builds:
  - id: "hcloud-build"
    main: ./cmd/hcloud/main.go
    binary: hcloud
    ldflags:
      - -w -X github.com/hetznercloud/cli/cli.Version={{.Version}}
    env:
      - CGO_ENABLED=0
    goos:
      - freebsd
      - windows
      - linux
    goarch:
      - amd64
      - arm
      - arm64
      - 386
    goarm:
      - 6
      - 7

  - id: "hcloud-macos-build"
    main: ./cmd/hcloud/main.go
    binary: hcloud
    ldflags:
      - -w -X github.com/hetznercloud/cli/cli.Version={{.Version}}
    env:
      - CGO_ENABLED=0
    goos:
      - darwin
    goarch:
      - amd64

before:
  hooks:
    - go mod tidy

release:
  draft: true
  name_template: "hcloud v{{.Version}}"
  disable: true

signs:
  - artifacts: all
    ids:
      - hcloud-build
    args: ["--batch", "-u", "hetzner-cloud-bot_test@lukas-kaemmerling.de", "--pinentry-mode", "loopback", "--output", "${signature}", "--detach-sign", "${artifact}"]
  - signature: "./dist/${artifact}.dmg"
    ids:
      - hcloud-macos-build
    cmd: gon
    args:
      - -log-json
      - gon.hcl
    artifacts: all

archives:
  - id: hcloud-archive
    builds:
      - hcloud-build
    name_template: "{{ .Binary }}-{{ .Os }}-{{ .Arch }}{{ if .Arm }}v{{ .Arm }}{{ end }}"

    format_overrides:
      - goos: windows
        format: zip

    files:
      - LICENSE
      - CHANGES.md
      - README.md

```

> Make sure your config file is valid by running
> `goreleaser check -f path-to-config-file`.

**Expected behavior**

A clear and concise description of what you expected to happen.

**Environment (please complete the following information):**

- OS: mac
- OS version: Catalina
- GoReleaser Version 0.124.1
Sample Run from Github Actions: https://github.com/LKaemmerling/cli/runs/412989126

**Additional context**
Stacktrace:
```
2020-01-28T14:11:28.5750190Z [command]/Users/runner/runners/2.164.0/work/_temp/dc3cc825-d08e-4e62-b747-65e066e2b811/goreleaser release --rm-dist --snapshot
2020-01-28T14:11:28.8862920Z 
2020-01-28T14:11:28.8866050Z    â€¢ releasing using goreleaser 0.124.1...
2020-01-28T14:11:28.8868890Z    â€¢ loading config file       file=.goreleaser.yml
2020-01-28T14:11:28.8872430Z    â€¢ RUNNING BEFORE HOOKS     
2020-01-28T14:11:28.8875960Z       â€¢ running go mod tidy      
2020-01-28T14:11:41.7605290Z    â€¢ LOADING ENVIRONMENT VARIABLES
2020-01-28T14:11:41.7640040Z       â€¢ pipe skipped              error=publishing is disabled
2020-01-28T14:11:41.7680040Z    â€¢ GETTING AND VALIDATING GIT STATE
2020-01-28T14:11:41.8283870Z       â€¢ releasing v99.99.5, commit 831d1b6c34019ea087ccaac488a5d37f83a45585
2020-01-28T14:11:41.8285720Z       â€¢ pipe skipped              error=disabled during snapshot mode
2020-01-28T14:11:41.8287410Z    â€¢ PARSING TAG              
2020-01-28T14:11:41.8288750Z    â€¢ SETTING DEFAULTS         
2020-01-28T14:11:41.8290530Z       â€¢ LOADING ENVIRONMENT VARIABLES
2020-01-28T14:11:41.8292110Z       â€¢ SNAPSHOTING              
2020-01-28T14:11:41.8293600Z       â€¢ GITHUB/GITLAB/GITEA RELEASES
2020-01-28T14:11:41.8394750Z       â€¢ PROJECT NAME             
2020-01-28T14:11:41.8396330Z       â€¢ BUILDING BINARIES        
2020-01-28T14:11:41.8398930Z       â€¢ ARCHIVES                 
2020-01-28T14:11:41.8400510Z       â€¢ LINUX PACKAGES WITH NFPM 
2020-01-28T14:11:41.8402180Z       â€¢ SNAPCRAFT PACKAGES       
2020-01-28T14:11:41.8403580Z       â€¢ CALCULATING CHECKSUMS    
2020-01-28T14:11:41.8404950Z       â€¢ SIGNING ARTIFACTS        
2020-01-28T14:11:41.8406300Z       â€¢ DOCKER IMAGES            
2020-01-28T14:11:41.8407660Z       â€¢ ARTIFACTORY              
2020-01-28T14:11:41.8409020Z       â€¢ S3                       
2020-01-28T14:11:41.8410920Z       â€¢ BLOB                     
2020-01-28T14:11:41.8412280Z       â€¢ HOMEBREW TAP FORMULA     
2020-01-28T14:11:41.8413760Z          â€¢ optimistically guessing `brew[0].installs`, double check
2020-01-28T14:11:41.8415810Z       â€¢ SCOOP MANIFEST           
2020-01-28T14:11:41.8417240Z    â€¢ SNAPSHOTING              
2020-01-28T14:11:41.8418580Z    â€¢ CHECKING ./DIST          
2020-01-28T14:11:41.8419930Z    â€¢ WRITING EFFECTIVE CONFIG FILE
2020-01-28T14:11:41.8421410Z       â€¢ writing                   config=dist/config.yaml
2020-01-28T14:11:41.8459450Z    â€¢ GENERATING CHANGELOG     
2020-01-28T14:11:41.8461070Z       â€¢ pipe skipped              error=not available for snapshots
2020-01-28T14:11:41.8462590Z    â€¢ BUILDING BINARIES        
2020-01-28T14:11:41.8463880Z       â€¢ building                  binary=dist/hcloud_linux_386/hcloud
2020-01-28T14:11:41.8465410Z       â€¢ building                  binary=dist/hcloud_freebsd_arm_7/hcloud
2020-01-28T14:11:41.8466860Z       â€¢ building                  binary=dist/hcloud_linux_amd64/hcloud
2020-01-28T14:11:41.8479580Z       â€¢ building                  binary=dist/hcloud_windows_amd64/hcloud.exe
2020-01-28T14:12:28.2502150Z       â€¢ building                  binary=dist/hcloud_freebsd_386/hcloud
2020-01-28T14:12:28.2872320Z       â€¢ building                  binary=dist/hcloud_windows_386/hcloud.exe
2020-01-28T14:12:28.5525050Z       â€¢ building                  binary=dist/hcloud_linux_arm_7/hcloud
2020-01-28T14:12:29.7933740Z       â€¢ building                  binary=dist/hcloud_linux_arm_6/hcloud
2020-01-28T14:13:10.1161370Z       â€¢ building                  binary=dist/hcloud_linux_arm64/hcloud
2020-01-28T14:13:10.2333040Z       â€¢ building                  binary=dist/hcloud_freebsd_amd64/hcloud
2020-01-28T14:13:11.4503160Z       â€¢ building                  binary=dist/hcloud_freebsd_arm_6/hcloud
2020-01-28T14:13:46.1441060Z       â€¢ building                  binary=dist/hcloud-macos_darwin_amd64/hcloud
2020-01-28T14:13:55.6005240Z    â€¢ ARCHIVES                 
2020-01-28T14:13:55.6017860Z       â€¢ creating                  archive=dist/hcloud-freebsd-armv6.tar.gz
2020-01-28T14:13:55.6019430Z       â€¢ creating                  archive=dist/hcloud-linux-armv7.tar.gz
2020-01-28T14:13:55.6020960Z       â€¢ creating                  archive=dist/hcloud-freebsd-386.tar.gz
2020-01-28T14:13:55.6022420Z       â€¢ creating                  archive=dist/hcloud-windows-amd64.zip
2020-01-28T14:13:56.5731260Z       â€¢ creating                  archive=dist/hcloud-linux-arm64.tar.gz
2020-01-28T14:13:56.5884840Z       â€¢ creating                  archive=dist/hcloud-freebsd-amd64.tar.gz
2020-01-28T14:13:56.6101590Z       â€¢ creating                  archive=dist/hcloud-linux-386.tar.gz
2020-01-28T14:13:56.6180010Z       â€¢ creating                  archive=dist/hcloud-linux-amd64.tar.gz
2020-01-28T14:13:57.9439250Z       â€¢ creating                  archive=dist/hcloud-linux-armv6.tar.gz
2020-01-28T14:13:57.9857470Z       â€¢ creating                  archive=dist/hcloud-freebsd-armv7.tar.gz
2020-01-28T14:13:58.2931230Z       â€¢ creating                  archive=dist/hcloud-windows-386.zip
2020-01-28T14:13:59.1867030Z    â€¢ LINUX PACKAGES WITH NFPM 
2020-01-28T14:13:59.1868320Z    â€¢ SNAPCRAFT PACKAGES       
2020-01-28T14:13:59.1869510Z    â€¢ CALCULATING CHECKSUMS    
2020-01-28T14:13:59.1973930Z       â€¢ checksumming              file=hcloud-linux-armv6.tar.gz
2020-01-28T14:13:59.2068720Z       â€¢ checksumming              file=hcloud-linux-armv7.tar.gz
2020-01-28T14:13:59.2077180Z       â€¢ checksumming              file=hcloud-linux-386.tar.gz
2020-01-28T14:13:59.2179790Z       â€¢ checksumming              file=hcloud-windows-amd64.zip
2020-01-28T14:13:59.2273160Z       â€¢ checksumming              file=hcloud-freebsd-amd64.tar.gz
2020-01-28T14:13:59.2319850Z       â€¢ checksumming              file=hcloud-linux-arm64.tar.gz
2020-01-28T14:13:59.2322210Z       â€¢ checksumming              file=hcloud-linux-amd64.tar.gz
2020-01-28T14:13:59.2324260Z       â€¢ checksumming              file=hcloud-windows-386.zip
2020-01-28T14:13:59.2326770Z       â€¢ checksumming              file=hcloud-freebsd-armv7.tar.gz
2020-01-28T14:13:59.2328690Z       â€¢ checksumming              file=hcloud-freebsd-386.tar.gz
2020-01-28T14:13:59.2333150Z       â€¢ checksumming              file=hcloud-freebsd-armv6.tar.gz
2020-01-28T14:13:59.2534540Z    â€¢ SIGNING ARTIFACTS        
2020-01-28T14:13:59.2535680Z fatal error: concurrent map writes
2020-01-28T14:13:59.2536570Z fatal error: concurrent map writes
2020-01-28T14:13:59.2981330Z 
2020-01-28T14:13:59.2982540Z goroutine 147 [running]:
2020-01-28T14:13:59.2983500Z runtime.throw(0x1f5fdf9, 0x15)
2020-01-28T14:13:59.2984760Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/runtime/panic.go:774 +0x72 fp=0xc00011db88 sp=0xc00011db58 pc=0x102d6a2
2020-01-28T14:13:59.2986780Z runtime.mapassign_faststr(0x1d81440, 0xc0003aa510, 0x1f528bc, 0x9, 0x28)
2020-01-28T14:13:59.2988180Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/runtime/map_faststr.go:291 +0x3fe fp=0xc00011dbf0 sp=0xc00011db88 pc=0x1012cbe
2020-01-28T14:13:59.2991870Z github.com/goreleaser/goreleaser/internal/pipe/sign.signone(0xc00028d600, 0xc000371068, 0x3, 0xc00036d580, 0x2, 0x2, 0xc00039cd40, 0x16, 0xc0003710b0, 0x3, ...)
2020-01-28T14:13:59.2994120Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/pipe/sign/sign.go:106 +0x127 fp=0xc00011dcf0 sp=0xc00011dbf0 pc=0x1b66fc7
2020-01-28T14:13:59.2995630Z github.com/goreleaser/goreleaser/internal/pipe/sign.sign(0xc00028d600, 0xc000371068, 0x3, 0xc00036d580, 0x2, 0x2, 0xc00039cd40, 0x16, 0xc0003710b0, 0x3, ...)
2020-01-28T14:13:59.2997200Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/pipe/sign/sign.go:94 +0xb8 fp=0xc00011dd90 sp=0xc00011dcf0 pc=0x1b66e38
2020-01-28T14:13:59.2998500Z github.com/goreleaser/goreleaser/internal/pipe/sign.Pipe.Run.func1(0xc000017700, 0x21a4693)
2020-01-28T14:13:59.2999800Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/pipe/sign/sign.go:86 +0x255 fp=0xc00011def0 sp=0xc00011dd90 pc=0x1b67995
2020-01-28T14:13:59.3001070Z github.com/goreleaser/goreleaser/internal/semerrgroup.(*parallelGroup).Go.func1(0x0, 0x0)
2020-01-28T14:13:59.3002370Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/semerrgroup/sem.go:43 +0x8d fp=0xc00011df58 sp=0xc00011def0 pc=0x118141d
2020-01-28T14:13:59.3003540Z golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0000b4008, 0xc00000c120)
2020-01-28T14:13:59.3005790Z 	/home/travis/gopath/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:57 +0x64 fp=0xc00011dfd0 sp=0xc00011df58 pc=0x1180fa4
2020-01-28T14:13:59.3007030Z runtime.goexit()
2020-01-28T14:13:59.3008090Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc00011dfd8 sp=0xc00011dfd0 pc=0x105ab61
2020-01-28T14:13:59.3009150Z created by golang.org/x/sync/errgroup.(*Group).Go
2020-01-28T14:13:59.3011800Z 	/home/travis/gopath/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:54 +0x66
2020-01-28T14:13:59.3012990Z 
2020-01-28T14:13:59.3014300Z goroutine 1 [select, 2 minutes]:
2020-01-28T14:13:59.3016150Z github.com/caarlos0/ctrlc.(*Ctrlc).Run(0xc000092270, 0x21f6ac0, 0xc00028d600, 0xc00038d0d0, 0x1, 0x1)
2020-01-28T14:13:59.3018190Z 	/home/travis/gopath/pkg/mod/github.com/caarlos0/ctrlc@v1.0.0/ctrlc.go:42 +0x1e0
2020-01-28T14:13:59.3020280Z main.releaseProject(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x100000001, 0x4, ...)
2020-01-28T14:13:59.3022190Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/main.go:153 +0x2d2
2020-01-28T14:13:59.3023830Z main.main()
2020-01-28T14:13:59.3025420Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/main.go:115 +0x1b5f
2020-01-28T14:13:59.3026020Z 
2020-01-28T14:13:59.3027000Z goroutine 18 [syscall, 2 minutes]:
2020-01-28T14:13:59.3044680Z os/signal.signal_recv(0x0)
2020-01-28T14:13:59.3046430Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/runtime/sigqueue.go:144 +0x96
2020-01-28T14:13:59.3048090Z os/signal.loop()
2020-01-28T14:13:59.3049050Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/os/signal/signal_unix.go:23 +0x22
2020-01-28T14:13:59.3050560Z created by os/signal.init.0
2020-01-28T14:13:59.3051540Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/os/signal/signal_unix.go:29 +0x41
2020-01-28T14:13:59.3052420Z 
2020-01-28T14:13:59.3053220Z goroutine 20 [select]:
2020-01-28T14:13:59.3054030Z go.opencensus.io/stats/view.(*worker).start(0xc0000bc370)
2020-01-28T14:13:59.3055100Z 	/home/travis/gopath/pkg/mod/go.opencensus.io@v0.22.0/stats/view/worker.go:154 +0x100
2020-01-28T14:13:59.3056100Z created by go.opencensus.io/stats/view.init.0
2020-01-28T14:13:59.3057080Z 	/home/travis/gopath/pkg/mod/go.opencensus.io@v0.22.0/stats/view/worker.go:32 +0x57
2020-01-28T14:13:59.3057660Z 
2020-01-28T14:13:59.3058750Z goroutine 21 [semacquire]:
2020-01-28T14:13:59.3060530Z sync.runtime_Semacquire(0xc0000b4018)
2020-01-28T14:13:59.3061520Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/runtime/sema.go:56 +0x42
2020-01-28T14:13:59.3062750Z sync.(*WaitGroup).Wait(0xc0000b4010)
2020-01-28T14:13:59.3063740Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/sync/waitgroup.go:130 +0x64
2020-01-28T14:13:59.3064860Z golang.org/x/sync/errgroup.(*Group).Wait(0xc0000b4008, 0xc0000b4008, 0xc00000c120)
2020-01-28T14:13:59.3066810Z 	/home/travis/gopath/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:40 +0x31
2020-01-28T14:13:59.3068160Z github.com/goreleaser/goreleaser/internal/semerrgroup.(*parallelGroup).Wait(0xc0000b4000, 0xc000116540, 0xc0001a0000)
2020-01-28T14:13:59.3069850Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/semerrgroup/sem.go:49 +0x31
2020-01-28T14:13:59.3072010Z github.com/goreleaser/goreleaser/internal/pipe/sign.Pipe.Run(0xc00028d600, 0x0, 0x2a9bfd0)
2020-01-28T14:13:59.3074150Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/pipe/sign/sign.go:89 +0x1c8
2020-01-28T14:13:59.3076580Z github.com/goreleaser/goreleaser/internal/middleware.ErrHandler.func1(0xc00028d600, 0xc0003c8180, 0x11)
2020-01-28T14:13:59.3078670Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/middleware/error.go:13 +0x33
2020-01-28T14:13:59.3079880Z github.com/goreleaser/goreleaser/internal/middleware.Logging.func1(0xc00028d600, 0x0, 0x0)
2020-01-28T14:13:59.3081210Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/middleware/logging.go:35 +0x24a
2020-01-28T14:13:59.3082230Z main.releaseProject.func1(0x0, 0x0)
2020-01-28T14:13:59.3083170Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/main.go:159 +0x11b
2020-01-28T14:13:59.3084250Z github.com/caarlos0/ctrlc.(*Ctrlc).Run.func1(0xc000092270, 0xc00038d0d0)
2020-01-28T14:13:59.3085380Z 	/home/travis/gopath/pkg/mod/github.com/caarlos0/ctrlc@v1.0.0/ctrlc.go:39 +0x27
2020-01-28T14:13:59.3087110Z created by github.com/caarlos0/ctrlc.(*Ctrlc).Run
2020-01-28T14:13:59.3088790Z 	/home/travis/gopath/pkg/mod/github.com/caarlos0/ctrlc@v1.0.0/ctrlc.go:38 +0x64
2020-01-28T14:13:59.3089810Z 
2020-01-28T14:13:59.3091110Z goroutine 146 [running]:
2020-01-28T14:13:59.3092500Z 	goroutine running on other thread; stack unavailable
2020-01-28T14:13:59.3094060Z created by golang.org/x/sync/errgroup.(*Group).Go
2020-01-28T14:13:59.3096860Z 	/home/travis/gopath/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:54 +0x66
2020-01-28T14:13:59.3486540Z 
2020-01-28T14:13:59.3487630Z goroutine 146 [running]:
2020-01-28T14:13:59.3488430Z runtime.throw(0x1f5fdf9, 0x15)
2020-01-28T14:13:59.3489940Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/runtime/panic.go:774 +0x72 fp=0xc00011eb88 sp=0xc00011eb58 pc=0x102d6a2
2020-01-28T14:13:59.3491140Z runtime.mapassign_faststr(0x1d81440, 0xc0003aa510, 0x1f528bc, 0x9, 0x21)
2020-01-28T14:13:59.3492410Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/runtime/map_faststr.go:291 +0x3fe fp=0xc00011ebf0 sp=0xc00011eb88 pc=0x1012cbe
2020-01-28T14:13:59.3493860Z github.com/goreleaser/goreleaser/internal/pipe/sign.signone(0xc00028d600, 0x1f4c7f1, 0x3, 0xc000329d40, 0x9, 0x9, 0x1f57ab5, 0xf, 0xc000370fb8, 0x3, ...)
2020-01-28T14:13:59.3496960Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/pipe/sign/sign.go:106 +0x127 fp=0xc00011ecf0 sp=0xc00011ebf0 pc=0x1b66fc7
2020-01-28T14:13:59.3500060Z github.com/goreleaser/goreleaser/internal/pipe/sign.sign(0xc00028d600, 0x1f4c7f1, 0x3, 0xc000329d40, 0x9, 0x9, 0x1f57ab5, 0xf, 0xc000370fb8, 0x3, ...)
2020-01-28T14:13:59.3502660Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/pipe/sign/sign.go:94 +0xb8 fp=0xc00011ed90 sp=0xc00011ecf0 pc=0x1b66e38
2020-01-28T14:13:59.3504990Z github.com/goreleaser/goreleaser/internal/pipe/sign.Pipe.Run.func1(0xc000019700, 0x21a4693)
2020-01-28T14:13:59.3507020Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/pipe/sign/sign.go:86 +0x255 fp=0xc00011eef0 sp=0xc00011ed90 pc=0x1b67995
2020-01-28T14:13:59.3508310Z github.com/goreleaser/goreleaser/internal/semerrgroup.(*parallelGroup).Go.func1(0x0, 0x0)
2020-01-28T14:13:59.3509670Z 	/home/travis/gopath/src/github.com/goreleaser/goreleaser/internal/semerrgroup/sem.go:43 +0x8d fp=0xc00011ef58 sp=0xc00011eef0 pc=0x118141d
2020-01-28T14:13:59.3511800Z golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0000b4008, 0xc00000c100)
2020-01-28T14:13:59.3515440Z 	/home/travis/gopath/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:57 +0x64 fp=0xc00011efd0 sp=0xc00011ef58 pc=0x1180fa4
2020-01-28T14:13:59.3517300Z runtime.goexit()
2020-01-28T14:13:59.3519110Z 	/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc00011efd8 sp=0xc00011efd0 pc=0x105ab61
2020-01-28T14:13:59.3520980Z created by golang.org/x/sync/errgroup.(*Group).Go
2020-01-28T14:13:59.3523720Z 	/home/travis/gopath/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:54 +0x66
2020-01-28T14:13:59.3531950Z ##[error]The process '/Users/runner/runners/2.164.0/work/_temp/dc3cc825-d08e-4e62-b747-65e066e2b811/goreleaser' failed with exit code 2
2020-01-28T14:13:59.3585070Z ##[error]Node run failed with exit code 1
2020-01-28T14:13:59.3784870Z Cleaning up orphan processes
2020-01-28T14:13:59.6637800Z Terminate orphan process: pid (1689) (gpg-agent)
```

--- a/internal/pipe/sign/sign.go
+++ b/internal/pipe/sign/sign.go
@@ -8,9 +8,9 @@ import (
 	"reflect"
 
 	"github.com/apex/log"
-
 	"github.com/goreleaser/goreleaser/internal/artifact"
 	"github.com/goreleaser/goreleaser/internal/deprecate"
+	"github.com/goreleaser/goreleaser/internal/ids"
 	"github.com/goreleaser/goreleaser/internal/logext"
 	"github.com/goreleaser/goreleaser/internal/pipe"
 	"github.com/goreleaser/goreleaser/internal/semerrgroup"
@@ -33,6 +33,7 @@ func (Pipe) Default(ctx *context.Context) error {
 			deprecate.Notice("sign")
 		}
 	}
+	var ids = ids.New("signs")
 	for i := range ctx.Config.Signs {
 		cfg := &ctx.Config.Signs[i]
 		if cfg.Cmd == "" {
@@ -47,8 +48,12 @@ func (Pipe) Default(ctx *context.Context) error {
 		if cfg.Artifacts == "" {
 			cfg.Artifacts = "none"
 		}
+		if cfg.ID == "" {
+			cfg.ID = "default"
+		}
+		ids.Inc(cfg.ID)
 	}
-	return nil
+	return ids.Validate()
 }
 
 // Run executes the Pipe.
@@ -101,7 +106,7 @@ func sign(ctx *context.Context, cfg config.Sign, artifacts []*artifact.Artifact)
 }
 
 func signone(ctx *context.Context, cfg config.Sign, a *artifact.Artifact) (*artifact.Artifact, error) {
-	env := ctx.Env
+	env := ctx.Env.Copy()
 	env["artifact"] = a.Path
 	env["signature"] = expand(cfg.Signature, env)
 
@@ -133,6 +138,9 @@ func signone(ctx *context.Context, cfg config.Sign, a *artifact.Artifact) (*arti
 		Type: artifact.Signature,
 		Name: name,
 		Path: filepath.Join(artifactPathBase, sigFilename),
+		Extra: map[string]interface{}{
+			"ID": cfg.ID,
+		},
 	}, nil
 }
 

--- a/internal/pipe/sign/sign_test.go
+++ b/internal/pipe/sign/sign_test.go
@@ -235,7 +235,7 @@ func testSign(t *testing.T, ctx *context.Context, signaturePaths []string, signa
 	// create temp dir for file and signature
 	tmpdir, err := ioutil.TempDir("", "goreleaser")
 	assert.NoError(t, err)
-	// defer os.RemoveAll(tmpdir)
+	defer os.RemoveAll(tmpdir)
 
 	ctx.Config.Dist = tmpdir
 
@@ -304,7 +304,6 @@ func testSign(t *testing.T, ctx *context.Context, signaturePaths []string, signa
 
 	// run the pipeline
 	if expectedErrMsg != "" {
-		// assert.
 		assert.EqualError(t, Pipe{}.Run(ctx), expectedErrMsg)
 		return
 	}

--- a/internal/pipe/sign/sign_test.go
+++ b/internal/pipe/sign/sign_test.go
@@ -9,6 +9,7 @@ import (
 	"os/exec"
 	"path/filepath"
 	"sort"
+	"strings"
 	"testing"
 	"time"
 
@@ -16,6 +17,7 @@ import (
 	"github.com/goreleaser/goreleaser/pkg/config"
 	"github.com/goreleaser/goreleaser/pkg/context"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 var originKeyring = "testdata/gnupg"
@@ -106,6 +108,36 @@ func TestSignArtifacts(t *testing.T) {
 			signatureNames: []string{"artifact1.sig", "artifact2.sig", "artifact3_1.0.0_linux_amd64.sig", "checksum.sig", "checksum2.sig", "artifact4_1.0.0_linux_amd64.sig"},
 		},
 		{
+			desc: "multiple sign configs",
+			ctx: context.New(
+				config.Project{
+					Signs: []config.Sign{
+						{
+							ID:        "s1",
+							Artifacts: "checksum",
+						},
+						{
+							ID:        "s2",
+							Artifacts: "checksum",
+							Signature: "${artifact}.sog",
+						},
+					},
+				},
+			),
+			signaturePaths: []string{
+				"checksum.sig",
+				"checksum2.sig",
+				"checksum.sog",
+				"checksum2.sog",
+			},
+			signatureNames: []string{
+				"checksum.sig",
+				"checksum2.sig",
+				"checksum.sog",
+				"checksum2.sog",
+			},
+		},
+		{
 			desc: "sign filtered artifacts",
 			ctx: context.New(
 				config.Project{
@@ -189,7 +221,7 @@ func testSign(t *testing.T, ctx *context.Context, signaturePaths []string, signa
 	// create temp dir for file and signature
 	tmpdir, err := ioutil.TempDir("", "goreleaser")
 	assert.NoError(t, err)
-	defer os.RemoveAll(tmpdir)
+	// defer os.RemoveAll(tmpdir)
 
 	ctx.Config.Dist = tmpdir
 
@@ -249,7 +281,12 @@ func testSign(t *testing.T, ctx *context.Context, signaturePaths []string, signa
 	// configure the pipeline
 	// make sure we are using the test keyring
 	assert.NoError(t, Pipe{}.Default(ctx))
-	ctx.Config.Signs[0].Args = append([]string{"--homedir", keyring}, ctx.Config.Signs[0].Args...)
+	for i := range ctx.Config.Signs {
+		ctx.Config.Signs[i].Args = append(
+			[]string{"--homedir", keyring},
+			ctx.Config.Signs[i].Args...,
+		)
+	}
 
 	// run the pipeline
 	assert.NoError(t, Pipe{}.Run(ctx))
@@ -276,7 +313,7 @@ func testSign(t *testing.T, ctx *context.Context, signaturePaths []string, signa
 
 	wantFiles := append(artifacts, signaturePaths...)
 	sort.Strings(wantFiles)
-	assert.Equal(t, wantFiles, gotFiles)
+	assert.ElementsMatch(t, wantFiles, gotFiles)
 
 	// verify the signatures
 	for _, sig := range signaturePaths {
@@ -288,11 +325,11 @@ func testSign(t *testing.T, ctx *context.Context, signaturePaths []string, signa
 		signArtifacts = append(signArtifacts, sig.Name)
 	}
 	// check signature is an artifact
-	assert.Equal(t, signArtifacts, signatureNames)
+	assert.ElementsMatch(t, signArtifacts, signatureNames)
 }
 
 func verifySignature(t *testing.T, ctx *context.Context, sig string) {
-	artifact := sig[:len(sig)-len(".sig")]
+	artifact := strings.Replace(sig, filepath.Ext(sig), "", 1)
 
 	// verify signature was made with key for usesr 'nopass'
 	cmd := exec.Command("gpg", "--homedir", keyring, "--verify", filepath.Join(ctx.Config.Dist, sig), filepath.Join(ctx.Config.Dist, artifact))
@@ -304,6 +341,22 @@ func verifySignature(t *testing.T, ctx *context.Context, sig string) {
 	// keyring before we do the verification. For now we punt and look in the
 	// output.
 	if !bytes.Contains(out, []byte(user)) {
-		t.Fatalf("signature is not from %s", user)
+		t.Fatalf("%s: signature is not from %s: %s", sig, user, string(out))
 	}
+}
+
+func TestSeveralSignsWithTheSameID(t *testing.T) {
+	var ctx = &context.Context{
+		Config: config.Project{
+			Signs: []config.Sign{
+				{
+					ID: "a",
+				},
+				{
+					ID: "a",
+				},
+			},
+		},
+	}
+	require.EqualError(t, Pipe{}.Default(ctx), "found 2 signs with the ID 'a', please fix your config")
 }

--- a/internal/pipe/sign/sign_test.go
+++ b/internal/pipe/sign/sign_test.go
@@ -80,7 +80,21 @@ func TestSignArtifacts(t *testing.T) {
 		ctx            *context.Context
 		signaturePaths []string
 		signatureNames []string
+		expectedErrMsg string
 	}{
+		{
+			desc:           "sign errors",
+			expectedErrMsg: "sign: exit failed",
+			ctx: context.New(
+				config.Project{
+					Sign: config.Sign{
+						Artifacts: "all",
+						Cmd:       "exit",
+						Args:      []string{"1"},
+					},
+				},
+			),
+		},
 		{
 			desc: "sign single",
 			ctx: context.New(
@@ -210,14 +224,14 @@ func TestSignArtifacts(t *testing.T) {
 
 	for _, test := range tests {
 		t.Run(test.desc, func(tt *testing.T) {
-			testSign(tt, test.ctx, test.signaturePaths, test.signatureNames)
+			testSign(tt, test.ctx, test.signaturePaths, test.signatureNames, test.expectedErrMsg)
 		})
 	}
 }
 
 const user = "nopass"
 
-func testSign(t *testing.T, ctx *context.Context, signaturePaths []string, signatureNames []string) {
+func testSign(t *testing.T, ctx *context.Context, signaturePaths []string, signatureNames []string, expectedErrMsg string) {
 	// create temp dir for file and signature
 	tmpdir, err := ioutil.TempDir("", "goreleaser")
 	assert.NoError(t, err)
@@ -289,7 +303,18 @@ func testSign(t *testing.T, ctx *context.Context, signaturePaths []string, signa
 	}
 
 	// run the pipeline
+	if expectedErrMsg != "" {
+		// assert.
+		assert.EqualError(t, Pipe{}.Run(ctx), expectedErrMsg)
+		return
+	}
+
 	assert.NoError(t, Pipe{}.Run(ctx))
+
+	// ensure all artifacts have an ID
+	for _, arti := range ctx.Artifacts.Filter(artifact.ByType(artifact.Signature)).List() {
+		assert.NotEmptyf(t, arti.ExtraOr("ID", ""), ".Extra.ID on %s", arti.Path)
+	}
 
 	// verify that only the artifacts and the signatures are in the dist dir
 	gotFiles := []string{}

--- a/pkg/config/config.go
+++ b/pkg/config/config.go
@@ -230,6 +230,7 @@ type NFPMOverridables struct {
 
 // Sign config
 type Sign struct {
+	ID        string   `yaml:"id,omitempty"`
 	Cmd       string   `yaml:"cmd,omitempty"`
 	Args      []string `yaml:"args,omitempty"`
 	Signature string   `yaml:"signature,omitempty"`

--- a/pkg/context/context.go
+++ b/pkg/context/context.go
@@ -28,6 +28,15 @@ type GitInfo struct {
 // Env is the environment variables
 type Env map[string]string
 
+// Copy returns a copy of the environment.
+func (e Env) Copy() Env {
+	var out = Env{}
+	for k, v := range e {
+		out[k] = v
+	}
+	return out
+}
+
 // Strings returns the current environment as a list of strings, suitable for
 // os executions.
 func (e Env) Strings() []string {

--- a/www/content/sign.md
+++ b/www/content/sign.md
@@ -28,6 +28,10 @@ To customize the signing pipeline you can use the following options:
 # .goreleaser.yml
 signs:
   -
+    # ID of the sign config, must be unique.
+    # Defaults to "default".
+    id: foo
+
     # name of the signature file.
     # '${artifact}' is the path to the artifact that should be signed.
     #

